### PR TITLE
[TCFMM-21] Fix price move in `y_to_x`

### PR DESCRIPTION
**Pull request by dcastro**
## Description
Problem: The `sqrt_price_move` functoin expects to be called with a `dx`
parameter.

However, in `y_to_x_rec`, it is called with `dy` instead of `dx`.

Solution: Write a function that computes the price move from `dy` (see
equation 6.13 from the whitepaper).

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

Resolves TCFMM-21

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

[//]: # (Update link to style guide if necesary or remove if it's not present)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by Martoon-00** [comment]
Minor, typo in the commit description:

>  The `sqrt_price_move` functoin expects to be called with a `dx` parameter.

File: ligo/helpers.mligo [hunk 1]
```diff
@@ -8,11 +8,69 @@
...
+
+        And indeed:
+            $ ligo compile-expression --init-file ligo/helpers.mligo cameligo \
+              "sqrt_price_move_y 100n {x80 = 12089258196146291747061760n} 1000n"
+            24178516392292583494123520
+   *)
+let sqrt_price_move_y (liquidity : nat) (sqrt_price_old : x80n) (dy : nat) : x80n =
+    (* ceildiv because we want to overstate how much this trade increases the price *)
+    { x80 =
+        Bitwise.shift_left (ceildiv dy liquidity) 80n + sqrt_price_old.x80
```
Comment (Martoon-00) [inline ligo/helpers.mligo:72]: Probably here you should apply `shift_left` in the first place, to decrease the precision loss in the division.

File: ligo/helpers.mligo [hunk 2]
```diff
@@ -8,11 +8,69 @@
...
+        Adding 10 `x` tokens to the pool should result in:
+            x = 20
+            y = L^2 / x = 500
+            P = 500 / 20 = 25
+            sqrt_price = sqrt(25) * 2^80 = 6044629098073145873530880
+
+        And indeed:
+            $ ligo compile-expression --init-file ligo/helpers.mligo cameligo \
+              "sqrt_price_move_x 100n {x80 = 12089258196146291747061760n} 10n"
+            6044629098073145873530880
```
Comment (Martoon-00) [inline ligo/helpers.mligo:35]: Nice writeup :+1:

File: ligo/helpers.mligo [hunk 3]
```diff
@@ -8,11 +8,69 @@
 #include "errors.mligo"
 #include "math.mligo"
 
-let sqrt_price_move (liquidity : nat) (sqrt_price : x80n) (dx : nat) : x80n =
+(*  Calculate the new `sqrt_price` after a deposit of `dx` `x` tokens.
+    Derived from equation 6.15:
+        Δ(1 / √P) = Δx / L
```
Comment (Martoon-00) [inline ligo/helpers.mligo:13]: In the article, without the parentheses, this was really a bit confusing, good to see them here.

**Comment by Martoon-00** [comment]
Btw, while you are at it, one minor suggestion:

After computing new sqrt price, we also compute the diff between the new and old values, and use `assert_nat` to assert that the price changed in the right direction.

From how `sqrt_price_move_*` are explained and implemented, it is clear that this `assert_nat` is valid. If so, maybe make `sqrt_price_move_*` compute and additionally return this diff?

**Review by Martoon-00 — APPROVED** [review]

File: ligo/helpers.mligo (hunk 1 above)
Comment (dcastro) [reply ligo/helpers.mligo]: Great catch! :smile:  Done!

**Comment by dcastro** [comment]
> Btw, while you are at it, one minor suggestion:
> 
> After computing new sqrt price, we also compute the diff between the new and old values, and use `assert_nat` to assert that the price changed in the right direction.
> 
> From how `sqrt_price_move_*` are explained and implemented, it is clear that this `assert_nat` is valid. If so, maybe make `sqrt_price_move_*` compute and additionally return this diff?

I like that suggestion!

I'm thinking we should put these extra "sanity checks" behind a `#if DEBUG` flag. We'll compile the contract with `DEBUG` for our tests, and without it for mainnet. What do you think?

File: ligo/helpers.mligo (hunk 1 above)
Comment (dcastro) [reply ligo/helpers.mligo]: Tried `sqrt_price_move_y` with some odd values (e.g. `liquidity = 100, dy = 3`), and this change does indeed increase the accuracy! :tada:

**Comment by Martoon-00** [comment]
> I'm thinking we should put these extra "sanity checks" behind a #if DEBUG flag. We'll compile the contract with DEBUG for our tests, and without it for mainnet. What do you think?

That's a very good idea! So, e.g. `assert_nat`s can be replaced with `abs` when `DEBUG` flag is off. 

If something goes wrong in mainnet, we will be in trouble, but hopefully we will catch all the issue before that.

**Comment by dcastro** [comment]
> That's a very good idea! So, e.g. `assert_nat`s can be replaced with way shorter `abs` when `DEBUG` flag is off.
> 
> If something goes wrong in mainnet, we will be in trouble, but hopefully we will catch all the issue before that.

Hmm I think the existing `assert_nat`s can stay, they're actively useful, and we use their results in other calculations.

I was only referring to sanity checks that don't produce anything useful (i.e. we ignore their output), are only there for the sole purpose of testing, like this:

```
let sqrt_price_move_y (liquidity : nat) (sqrt_price_old : x80n) (dy : nat) : x80n =
    (* ceildiv because we want to overstate how much this trade increases the price *)
    let sqrt_price_new = ceildiv (Bitwise.shift_left dy 80n) liquidity + sqrt_price_old.x80 in
#if DEBUG
    let _ = if sqrt_price_new < sqrt_price_old.x80
                then unit
                else failwith "sqrt_price moved in the wrong direction" in
#endif
    { x80 = sqrt_price_new }
```

**Comment by dcastro** [comment]
@Martoon-00 I've created a new ticket here: https://issues.serokell.io/issue/TCFMM-24

I'm gonna merge this PR if that's ok ^^

**Comment by Martoon-00** [comment]
Got a second look at your comment.

> Hmm I think the existing assert_nats can stay, they're actively useful, and we use their results in other calculations.

This way some `assert_nat`s will duplicate the sanity checks, won't they?

```
if sqrt_price_new < sqrt_price_old.x80
```

vs

```
let diff = assert_nat (sqrt_price_old.x80 - sqrt_price_new)
```

(I assume we don't care whether price _changed_, but care only about direction).

Will there be any value in the debug sanity checks, if we do the checking in either way and the thrown error is unique?

My point was to simplify judging about internal errors or even decrease the number of internal errors - by moving `assert_nat` into `move_sqrt_price_*`. Currently, errors in those `assert_nat` rely on the `move_sqrt_price_*` behaviour and that complicates the reasoning about the code.

**Pull request merged**